### ISSUE TITLE
베이스 라인 구축 - 성진님 개인과제

### ISSUE THREAD

**Issue by wynn3312** [SKNETWORKS-FAMILY-AICAMP/SKN27-2nd-2TEAM#13]
## 🛩️ Description
차주 월요일에 feature engineering을 진행하기 전 까지
기본적인 모델을 만들고 다룰 수 있는 기능을 키우는 것이 목표 입니다

만들어진 모델이 괜찮다면 해당 모델을 가지고 실험을 하면서 학습을 진행할 수도 있습니다.
구체적인 목표는 아래와 같습니다.

### 1. 강사님의 kaggle - Titanic 모델을 가지고 Netflex 데이터를 다룰 수 있는 기본 모델 제작
- 데이터 전처리나 인코딩 등 구체적인 사항들은 모두 다 배제하고 숫자형 데이터만 남긴 상태의 모델을 제작
- 해당 모델의 목표는 데이터를 분할 > 학습 가능한 데이터만 남김 > 모델 생성 > 학습 > 결과 평가 의 기본 처리 과정을 보는 겁니다.

### 2. 제작한 기본 모델에 결측치, 중복값 제거, 인코딩 처리 등을 적용해서 평가 점수를 상승시켜보자
- 기본적인 데이터 처리 과정에 대한 이해도를 확인하기 위함입니다.

### (선택) 3. colab 기준으로 만들어 졌던 모델을 내 PC (로컬 컴퓨터)에서 동작하도록 수정합니다.
colab에서 동작하도록 작업했던 파일을 내 PC에서도 동작시킬 수 있는지 확인하기 위한 작업 입니다, .
해당 작업은 코드를 다루는 전반적인 이해도를 확인하기 위한 부분으로 필수 사항은 아니고 선택 입니다

### (선택) 4. 정제된 데이터 셋이 아닌 실제 사용할 데이터 셋으로 교체해서 모델 분석을 진행합니다.
- 우리가 월요일에 실제 다룰 데이터는 데이터가 정제되지 않고 여러개의 테이블로 나눠져 있습니다.
- 해당 데이터들을 유저 ID 단위로 취합해서 하나의 데이터 테이블로 모아야 분석을 시작할 수 있습니다. 
- 데이터셋을 합치는 방법이나 어떻게 합쳐야 하는지 등은 GPT나 Gemini에게 물어보면서 진행하시면 됩니다.
- 결과물로 사용할 데이터 테이블이 csv 파일 하나로 나와야 합니다. 

## ☑️ Todos
- [x] 강사님 자료를 기초로 base model 제작 (해당 동작을 할 수 있는 기본 모델을 만들어 봅니다.
- [x] 넷플릭스 데이터에서 결측치, 중복값 제거, 인코딩 처리 등을 추가해 모델의 성능을 올려 봅니다.
- [x] (선택) colab 기준으로 만들어 졌던 모델을 내 PC (로컬 컴퓨터)에서 동작하도록 수정합니다.
- [x] (선택) 정제된 데이터 셋이 아닌 실제 사용할 데이터 셋으로 교체해서 모델 분석을 진행합니다.

## 🔗 Reference
(실제 작업데이터) 가상의 넷플릭스 데이터셋: 사용자, 영화, 시청 이력, 추천 기록, 검색 기록, 리뷰
https://www.kaggle.com/datasets/sayeeduddin/netflix-2025user-behavior-dataset-210k-records

(베이스라인 구축용 연습 데이터) 이탈률 예측을 위해 전처리된 상태로 제공되는 캐글 데이터셋
https://www.kaggle.com/datasets/rhythmghai/netflix-user-watching-behavior-dataset

**Comment by soiez2958** [comment]
# 📺 Netflix 고객 이탈 예측 프로젝트
 
## 📌 프로젝트 개요
 
본 프로젝트는 Kaggle에서 제공하는 Netflix 사용자 데이터를 활용하여 **고객 이탈(Churn) 예측 모델**을 구축하는 것을 목표로 진행되었습니다.  
팀 내 담당 역할은 **베이스 모델 구축 및 피처 생성, EDA 초반 작업**입니다.
 
---
 
## 📂 사용 데이터
 
Kaggle의 Netflix 데이터셋을 구성하는 총 6개의 CSV 파일을 활용하였습니다.
 
| 파일명 | 설명 |
|---|---|
| `users.csv` | 사용자 기본 정보 |
| `watch_history.csv` | 시청 기록 |
| `movies.csv` | 콘텐츠 정보 |
| `recommendation_logs.csv` | 추천 로그 |
| `search_logs.csv` | 검색 로그 |
| `reviews.csv` | 리뷰 데이터 |
 
---
 
## 🔧 피처 구성 (독립변수 10개)
 
6개의 원본 파일을 기반으로 총 **10개의 독립변수**를 구성하였습니다.
 
### 기존 변수 (6개)
 
| 변수명 | 설명 |
|---|---|
| `age` | 나이 |
| `gender` | 성별 |
| `country` | 국가 |
| `subscription_type` | 구독 유형 |
| `primary_device` | 주 사용 기기 |
| `monthly_fee` | 월 요금 |

## 🧹 데이터 전처리 (EDA)
 
### 결측치 처리
 
각 컬럼의 특성에 맞게 결측치를 처리하였습니다.
 
| 컬럼 | 처리 방법 |
|---|---|
| `age` | train 데이터의 중앙값으로 대체 |
| `gender` | 최빈값으로 대체 (`Male`, `Female`, `Other`, `Unknown` 4가지 값 존재) |
| `monthly_fee` | 연령대별 중앙값으로 대체 (10대는 0으로 처리) |
| `household_size` | 연령대별 중앙값으로 대체 |
| `completion_rate` / `cost_per_hour` | 시청 기록 없는 경우 0으로 처리 (분모 0 방지를 위해 보정값 0.1분 추가) |
 
> ✅ 처리 결과: train / test 모두 **결측치 0건** 완전 해소
 
### 이상치 탐지 (IQR 기반)
 
IQR(사분위범위) 방법으로 수치형 컬럼별 이상치를 탐지하고 박스플롯으로 시각화하였습니다.
 
| 컬럼 | 이상치 수 | 비율 |
|---|---|---|
| `devices_used` | 1,673개 | 20.91% |
| `avg_watch_time_minutes` | 423개 | 5.29% |
| `days_since_last_activity` | 389개 | 4.86% |
| `monthly_fee` | 379개 | 4.74% |
| `is_churned` (타겟) | 1,185개 | 14.81% |
 
> ⚠️ 타겟 변수 `is_churned`의 이상치 비율(14.81%)은 **클래스 불균형** 문제를 시사합니다.  
> 이탈(Churn) 비율이 약 15%에 불과하여, 추후 언더샘플링 또는 SMOTE 적용이 필요합니다.
 
### 인코딩
 
범주형 변수 5개에 대해 **OneHotEncoding**을 적용하였습니다.
 
- 인코딩 대상: `gender`, `subscription_type`, `primary_device`, `country`, `favorite_genre`
- 인코딩 전: 18개 컬럼 → 인코딩 후: **48개 컬럼**
 
---
 
## 🤖 모델 학습 및 평가
 
### 데이터 분할
 
전체 10,000건을 `stratify` 옵션을 적용하여 분할하였습니다.
 
- Train: **8,000건**
- Test: **2,000건**

### 모델 비교
 
DecisionTree를 베이스 모델로 설정하고, XGBoost / LightGBM / CatBoost와 성능을 비교하였습니다.
 
| 모델 | Train Score | Test Score | AUC | 과적합 여부 |
|---|---|---|---|---|
| DecisionTree | 0.9480 | 0.7190 | 0.5001 | ⚠️ 과적합 의심 |
| XGBoost | 0.9461 | 0.8435 | 0.5072 | ⚠️ 과적합 의심 |
| LightGBM | 0.8751 | 0.8505 | 0.4881 | ✅ 양호 |
| CatBoost | 0.8668 | 0.8520 | 0.4807 | ✅ 양호 |
 
> 🏆 AUC 기준 최고 성능 모델인 **XGBoost**를 최종 모델로 선정하였습니다.
 
### 최종 모델 평가 결과 (XGBoost)
 
| 지표 | 값 |
|---|---|
| Accuracy | 0.8435 |
| Loss | 0.1565 |
| Precision | 0.1600 |
| Recall | 0.0135 |
| F1 Score | 0.0249 |
| AUC | 0.5072 |

### 혼동 행렬 (Confusion Matrix)
 
| | 예측: 유지 | 예측: 이탈 |
|---|---|---|
| **실제: 유지** | 1,683명 (TN) | 21명 (FP) |
| **실제: 이탈** | 292명 (FN) | 4명 (TP) |
 
> ❗ Accuracy는 84.35%로 높지만, Recall이 1.35%에 불과합니다.  
> 실제 이탈 고객 296명 중 단 **4명**만 정확히 예측한 것으로, 클래스 불균형이 성능에 직접적인 영향을 미쳤습니다.
 
---
 
## 🗂️ 코드 모듈 구조
 
주요 기능을 함수화하여 모듈로 분리하였습니다. 모든 모듈은 `model_netflix.py`의 `model_Netflix()` 함수에서 순서대로 호출됩니다.
 
```
📦 프로젝트 구조
├── model_netflix.py              # 메인 실행 파일
├── Data_Preprocessing/
│   ├── load_data.py              # 데이터 로드 및 피처 빌딩
│   ├── merge_data.py             # 데이터 병합
│   ├── data_split.py             # Train/Test 분리
│   ├── cleaning.py               # 결측치 처리
│   ├── outlier_handling.py       # 이상치 탐지 및 시각화
│   └── encoding.py               # OneHotEncoding
├── Modeling/
│   ├── Model_test.py             # 모델 학습 및 평가
│   ├── check.py                  # 학습 전 데이터 검증
│   └── save_model.py             # 모델 및 결과 저장
└── Display_graph/
    ├── cm_fi_graph.py            # 혼동행렬 및 피처 중요도 시각화
    └── ac_ls_graph.py            # Accuracy & Loss 시각화
```
 
| 모듈 | 파일 | 주요 함수 |
|---|---|---|
| 데이터 로드 | `load_data.py` | `get_args()`, `load_data()` |
| 데이터 분할 | `data_split.py` | `split_data()` |
| 결측치 처리 | `cleaning.py` | `check_data_info()`, `fill_missing_values()` |
| 이상치 시각화 | `outlier_handling.py` | `detect_outliers()`, `plot_outliers()` |
| 인코딩 | `encoding.py` | `get_encode_cols()`, `encode_features()` |
| 모델 학습/평가 | `Model_test.py` | `train_models()`, `evaluate_all_models()` |
| 혼동행렬/그래프 | `cm_fi_graph.py`, `ac_ls_graph.py` | `plot_confusion_matrix()`, `plot_accuracy_loss()` |
| 모델 저장 | `save_model.py` | `save_model()`, `save_results()` |
 
---
 
## 💬 회고
 
### 잘된 점
 
- 6개 파일 기반의 피처 엔지니어링과 파생변수 생성으로 의미 있는 독립변수를 확보하였습니다.
- 전체 파이프라인을 모듈 단위로 분리하여 코드의 가독성과 재사용성을 높였습니다.
 
### 아쉬운 점 및 개선 방향
 
#### ① 데이터 전처리 순서 문제
 
EDA는 결측치 처리 이전에 수행했어야 하는데, 처리 이후에 진행하여 분석 방향이 흔들리는 문제가 발생하였습니다.  
또한 컬럼 내 이상치를 충분히 확인하지 않고 평균값·최빈값으로 일괄 대체하여 데이터 품질이 저하되었습니다.
 
- `age` 컬럼에 100세 이상 값이 다수 존재했음에도 평균값으로 단순 대체
- `gender` 컬럼의 경우 4가지 값이 존재했지만 3가지로만 처리하여 진행
 
#### ② 클래스 불균형 미처리
 
이탈(15%) vs 유지(85%)의 심각한 불균형이 존재했음에도 별도 처리 없이 학습을 진행하여 Recall이 극히 낮은 결과가 나왔습니다.
 
> 개선 방안: **언더샘플링** 또는 **SMOTE** 적용 필요
 
#### ③ 유의미한 피처 부재
 
상관관계 분석(히트맵) 결과, 타겟 변수와의 선형 관계가 `-0.03 ~ 0.01` 수준으로 매우 미약하였습니다.  
`cost_per_hour`와 `monthly_fee` 간 높은 상관관계(0.83)는 공식 자체에서 기인한 것으로, 실질적인 예측력을 가진 피처 발굴이 추가로 필요합니다.

**Comment by soiez2958** [comment]
<img width="1500" height="688" alt="Image" src="https://github.com/user-attachments/assets/69351573-1a7c-49df-8251-add9b004d21a" />
<img width="1000" height="500" alt="Image" src="https://github.com/user-attachments/assets/d2b07228-76f2-4b2a-bf27-2fc903f6cd76" />
<img width="800" height="600" alt="Image" src="https://github.com/user-attachments/assets/d4998a47-6776-4b3c-af8c-90ccd467977d" />
<img width="800" height="700" alt="Image" src="https://github.com/user-attachments/assets/dfa6a7e4-e288-4004-ac8a-73d50180e105" />
